hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -901,16 +904,34 @@ func (r *NodePoolReconciler) delete(ctx context.Context, nodePool *hyperv1.NodeP
 		return fmt.Errorf("failed to delete ConfigMaps with nodePool label: %w", err)
 	}
 
-	if err := deleteMachineDeployment(ctx, r.Client, md); err != nil {
-		return fmt.Errorf("failed to delete MachineDeployment: %w", err)
+	if nodePool.Spec.Platform.AWS != nil {
```

Comment:
Please let's drop this code. We dont' want to do an specific platform check here, nor rely on sleep.

What are you trying to achieve with this logic?

Reply:
I was trying to prevent the NodePool secrets from being deleted before the machines are fully deleted. It’s related to this comment - https://github.com/openshift/hypershift/pull/1826#discussion_r1006055749

Reply:
I see, so please paste a link with the diff with the change in capa that requires to do this now.
Let's fix it upstream. And here for now let's gate on Machines being gone.

Reply:
This change seems to have come from [PR 2271](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271). Specifically from the [changes](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271/files#diff-4f8240f4e7ff07dcfa9627342b62772c1fe6ae0943021abe53fbb966db88886e) in machine.go in `UseSecretsManager ` and the [changes](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271/files#diff-0b559bbd149f0e6d54d789235423f66fa8906fdc6ee9c99b9e85db912912011e) in awsmachine_controller.go in `reconcileDelete`.

Reply:
Forgot to add an example of the log messages. The following errors are observed when the secret is deleted before the AWS machine has fully terminated:
```
I1019 13:05:07.838355 1 awsmachine_controller.go:287] "msg"="Handling deleted AWSMachine"
I1019 13:05:07.838456 1 awsmachine_controller.go:763] "msg"="Deleting unneeded entry from AWS S3" "secretPrefix"=""
I1019 13:05:07.838473 1 awsmachine_controller.go:287] "msg"="Handling deleted AWSMachine"
E1019 13:05:07.838499 1 awsmachine_controller.go:292] "msg"="unable to delete machine" "error"="failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-us-east-1a-566vx: Secret \"user-data-brcox-hypershift-x86-us-east-1a-5c8b7c39\" not found"
I1019 13:05:07.838557 1 awsmachine_controller.go:763] "msg"="Deleting unneeded entry from AWS S3" "secretPrefix"=""
E1019 13:05:07.838589 1 awsmachine_controller.go:292] "msg"="unable to delete machine" "error"="failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-us-east-1a-bsx4q: Secret \"user-data-brcox-hypershift-x86-us-east-1a-5c8b7c39\" not found"
E1019 13:05:07.839019 1 controller.go:317] controller/awsmachine "msg"="Reconciler error" "error"="failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-us-east-1a-566vx: Secret \"user-data-brcox-hypershift-x86-us-east-1a-5c8b7c39\" not found" "name"="brcox-hypershift-x86-us-east-1a-566vx" "namespace"="clusters-brcox-hypershift-x86" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AWSMachine"
I1019 13:05:07.839079 1 recorder.go:103] events "msg"="Warning" "message"="failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-us-east-1a-566vx: Secret \"user-data-brcox-hypershift-x86-us-east-1a-5c8b7c39\" not found" "object"={"kind":"AWSMachine","namespace":"clusters-brcox-hypershift-x86","name":"brcox-hypershift-x86-us-east-1a-566vx","uid":"c0014da8-985d-481d-92bf-d6d7fb290e4a","apiVersion":"infrastructure.cluster.x-k8s.io/v1beta1","resourceVersion":"8072491"} "reason"="FailedGetBootstrapData"
I1019 13:05:07.839119 1 recorder.go:103] events "msg"="Warning" "message"="failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-us-east-1a-566vx: Secret \"user-data-brcox-hypershift-x86-us-east-1a-5c8b7c39\" not found" "object"={"kind":"AWSMachine","namespace":"clusters-brcox-hypershift-x86","name":"brcox-hypershift-x86-us-east-1a-566vx","uid":"c0014da8-985d-481d-92bf-d6d7fb290e4a","apiVersion":"infrastructure.cluster.x-k8s.io/v1beta1","resourceVersion":"8072491"} "reason"="FailedGetBootstrapData"
...
```